File: src/services/editLockService.ts
import apiClient from './apiClient';

export interface EditLockDTO {
  fileId: string;
  userId: string;
  userName: string;
  lockedAt: string;
}

const editLockService = {
  /** 取得所有編輯鎖定 */
  getAll: async (): Promise<EditLockDTO[]> => {
    const { data } = await apiClient.get<EditLockDTO[]>('/edit-locks');
    return data;
  },

  /** 取得或鎖定檔案 */
  acquire: async (fileId: string): Promise<{ success: boolean; lock?: EditLockDTO }> => {
    const { data } = await apiClient.post(`/edit-locks/${fileId}`);
    return data;
  },

  /** 釋放鎖定 */
  release: async (fileId: string): Promise<void> => {
    await apiClient.delete(`/edit-locks/${fileId}`);
  },
};

export default editLockService;

File: backend/TaoyuanDMS.API/Services/EditLockService.cs
using Dapper;
using TaoyuanDMS.API.Models;

namespace TaoyuanDMS.API.Services;

public class EditLockService
{
    private readonly DbConnectionFactory _db;

    public EditLockService(DbConnectionFactory db) => _db = db;

    public async Task<List<EditLockDto>> GetAllAsync()
    {
        using var conn = _db.CreateConnection();
        var locks = await conn.QueryAsync<EditLockDto>("SELECT * FROM EditLocks");
        return locks.ToList();
    }

    public async Task<(bool Success, EditLockDto? Lock)> AcquireAsync(string fileId, string userId, string userName)
    {
        using var conn = _db.CreateConnection();
        var existing = await conn.QueryFirstOrDefaultAsync<EditLockDto>(
            "SELECT * FROM EditLocks WHERE FileId = @FileId", new { FileId = fileId });

        if (existing != null)
        {
            if (existing.UserId == userId) return (true, existing);
            return (false, existing);
        }

        var lockDto = new EditLockDto { FileId = fileId, UserId = userId, UserName = userName, LockedAt = DateTime.UtcNow.ToString("o") };
        await conn.ExecuteAsync(
            "INSERT INTO EditLocks (FileId, UserId, UserName) VALUES (@FileId, @UserId, @UserName)",
            new { FileId = fileId, UserId = userId, UserName = userName });

        return (true, lockDto);
    }

    public async Task ReleaseAsync(string fileId, string userId)
    {
        using var conn = _db.CreateConnection();
        await conn.ExecuteAsync("DELETE FROM EditLocks WHERE FileId = @FileId AND UserId = @UserId",
            new { FileId = fileId, UserId = userId });
    }
}
